### PR TITLE
storage: support reverse raw_scan and raw_batch_scan

### PR DESCRIPTION
**Pull request by kamijin-fanta**
## What have you changed? (mandatory)

- support to reverse scan
- to follow the update of kvproto. https://github.com/pingcap/kvproto/pull/304

## What are the type of the changes? (mandatory)

- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)

yes. I added unit test for raw_batch_scan.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

no.

## Does this PR affect tidb-ansible update? (mandatory)

no.

## Refer to a related PR or issue link (optional)

- #3705
- https://github.com/pingcap/kvproto/pull/304

### CODE REVIEW

**Comment by MyonKeminta** [comment]
Great job!
Please sign-off your commits and run `cargo fmt`

**Comment by MyonKeminta** [comment]
Yes! And, please also fix the error messages from CircleCI

File: src/storage/mod.rs [hunk 1]
```diff
@@ -1223,15 +1243,18 @@ impl<E: Engine> Storage<E> {
...
         for i in 0..ranges_len {
             let start_key = ranges[i].get_start_key();
             let mut end_key = ranges[i].get_end_key();
             if end_key.is_empty() && i + 1 != ranges_len {
                 end_key = ranges[i + 1].get_start_key();
             }
-            if !end_key.is_empty() && start_key >= end_key {
+            if !end_key.is_empty()
+                || !reverse && start_key >= end_key
+                || reverse && start_key <= end_key
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: It seems like the ranges are invalid when the end_key is not empty 🤔 
May be you want to write
```rust
if !end_key.is_empty()
    && (!reverse && start_key >= end_key)
    && (reverse && start_key <= end_key)
```

In fact, I suppose that when the `KeyRange` is used, the range it represents should be [start_key, end_key). I means, you don't need to change the function `check_key_ranges`.

File: src/storage/mod.rs [hunk 2]
```diff
@@ -2702,7 +2769,41 @@ mod tests {
...
+                let mut range = KeyRange::new();
+                range.set_start_key(s);
+                range.set_end_key(e);
+                range
+            })
+            .collect();
+        expect_multi_values(
+            results,
+            storage
+                .async_raw_batch_scan(Context::new(), "".to_string(), ranges, 5, true, true)
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: It's better if your tests covers both raw_scan and raw_batch_scan.

**Comment by MyonKeminta** [comment]
In fact, I think it's better if the range to scan is always [start_key, end_key) whenever we are scanning forward or backward, which is more close to our semantics of the term "start_key" and "end_key".
@breeswish @zhangjinpeng1987 PTAL

File: src/storage/mod.rs (hunk 1 above)
Comment (kamijin-fanta) [reply src/storage/mod.rs]: oh. I made a mistake.

File: src/storage/mod.rs (hunk 2 above)
Comment (kamijin-fanta) [reply src/storage/mod.rs]: thank you. I added.

**Comment by kamijin-fanta** [comment]
lowerBound / upperBound and startKey / endKey are distinguished in the code.
so, I wrote such a code.

**Comment by breeswish** [comment]
Actually currently our semantic is that there is a range [lower_bound, upper_bound). when we do forward scan, we scan from lower_bound to upper_bound in the range. When we do reverse scan, we scan from upper_bound to lower_bound in the range. Maybe we can rename start_key to lower_bound and end_key to upper_bound to be less ambiguous?

**Comment by MyonKeminta** [comment]
@breeswish Seems it's not easy, since you may need to change the fields' name in kvproto

**Comment by MyonKeminta** [comment]
@breeswish oh, sorry, it seems renaming fields in kvproto won't break compatibility between tikv and rpc client..?

**Comment by MyonKeminta** [comment]
At least I think the fields of `KeyRange` should keep their meanings.
Neither should we rename the fields in it.

In coprocessor.proto:
```proto
// [start, end)
message KeyRange {
    bytes start = 1;
    bytes end = 2;
}
```
In kvrpcpb.proto
```proto
message KeyRange {
  bytes start_key = 1;
  bytes end_key = 2;
}
```

If we change them, we may introduce too much changes to TiKV and TiDB.
I think it's better to let [start_key, end_key) equivalent to [lower_bound, upper_bound) now.

**Comment by kamijin-fanta** [comment]
the consistency between startKey of ScanRequest and startKey of inside is lost.
```proto
message ScanRequest {
    Context context = 1;
    bytes start_key = 2;
    uint32 limit = 3;
    uint64 version = 4;
    bool key_only = 5;
    bool reverse = 6;
}
```
https://github.com/pingcap/kvproto/blob/855d2192cdc79c7353cf36e6b60e049bf436dc8a/proto/kvrpcpb.proto#L94-L101

I think it is easy to understand clearly distinguish between startKey and lowerBound.

**Comment by MyonKeminta** [comment]
OK. Let's accept this. But please add some comments to explain why `start_key` is the greater one.

File: src/storage/mod.rs [hunk 3]
```diff
@@ -1134,14 +1134,28 @@ impl<E: Engine> Storage<E> {
...
+        let scan_mode = if reverse {
+            ScanMode::Backward
+        } else {
+            ScanMode::Forward
+        };
+        let mut cursor = snapshot.iter_cf(Self::rawkv_cf(cf)?, option, scan_mode)?;
         let statistics = statistics.mut_cf_statistics(cf);
-        if !cursor.seek(start_key, statistics)? {
+        if reverse {
+            if !cursor.seek_for_prev(start_key, statistics)? {
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: Our ranges are always left-close right-open and `seek_for_prev` is inclusive. Please consider using `reverse_seek` instead.

File: src/storage/mod.rs [hunk 4]
```diff
@@ -2491,29 +2515,62 @@ mod tests {
...
+                    b"c2".to_vec(),
+                    20,
+                    false,
+                    false,
+                )
+                .wait(),
+        );
+        let mut results: Vec<Option<KvPair>> =
+            test_data.into_iter().map(|(k, v)| Some((k, v))).collect();
+        results.reverse();
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: Why not:
```rust
let mut results: Vec<Option<KvPair>> =
    test_data.into_iter().map(|(k, v)| Some((k, v))).rev().collect();
```

File: src/storage/mod.rs [hunk 5]
```diff
@@ -2491,29 +2515,62 @@ mod tests {
...
+            results,
+            storage
+                .async_raw_scan(
+                    Context::new(),
+                    "".to_string(),
+                    b"z".to_vec(),
+                    20,
+                    false,
+                    true,
+                )
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: Please add more tests with different start_keys and end_keys to see how they works.

File: src/storage/mod.rs (hunk 3 above)
Comment (MyonKeminta) [reply src/storage/mod.rs]: Also notice that your tests also need to be updated.

File: src/storage/mod.rs (hunk 4 above)
Comment (kamijin-fanta) [reply src/storage/mod.rs]: I did not know. thank you.

File: src/storage/mod.rs (hunk 3 above)
Comment (kamijin-fanta) [reply src/storage/mod.rs]: fixed

File: src/storage/mod.rs (hunk 5 above)
Comment (kamijin-fanta) [reply src/storage/mod.rs]: i added

File: src/storage/mod.rs [hunk 6]
```diff
@@ -1223,15 +1243,20 @@ impl<E: Engine> Storage<E> {
         Err(Error::InvalidCf(cf.to_owned()))
     }
 
-    fn check_key_ranges(ranges: &[KeyRange]) -> bool {
+    fn check_key_ranges(ranges: &[KeyRange], reverse: bool) -> bool {
+        // note:
+        //   if reverse, endKey is less than startKey. endKey is lowerBound.
+        //   if not reverse, endKey is greater than startKey. endKey is upperBound
```
Comment (ice1000) [inline src/storage/mod.rs]: Can we make this note the Rustdoc for this function?

Comment (kamijin-fanta) [reply src/storage/mod.rs]: thanks! i fixed.

**Comment by ice1000** [comment]
```
test raftstore_cases::test_split_region::test_delay_split_region ... thread 'raftstore_cases::test_split_region::test_delay_split_region' panicked at 'can't get value None for key "k0"', components/test_raftstore/src/util.rs:64:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:475
   5: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:390
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:345
   7: test_raftstore::util::must_get
             at components/test_raftstore/src/util.rs:64
   8: test_raftstore::util::must_get_none
             at components/test_raftstore/src/util.rs:76
   9: integrations::raftstore_cases::test_split_region::check_cluster
             at tests/raftstore_cases/test_split_region.rs:279
  10: integrations::raftstore_cases::test_split_region::test_delay_split_region
             at tests/raftstore_cases/test_split_region.rs:335
  11: integrations::__test::TESTS::{{closure}}
             at tests/raftstore_cases/test_split_region.rs:288
  12: core::ops::function::FnOnce::call_once
             at /checkout/src/libcore/ops/function.rs:223
  13: <F as alloc::boxed::FnBox<A>>::call_box
             at libtest/lib.rs:1454
             at /checkout/src/libcore/ops/function.rs:223
             at /checkout/src/liballoc/boxed.rs:640
  14: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:106
FAILED
```

**Comment by ice1000** [comment]
Please fix test and rebase onto the master branch

File: src/storage/mod.rs [hunk 7]
```diff
@@ -2491,29 +2519,111 @@ mod tests {
...
+            (b"d".to_vec(), b"dd".to_vec()),
+            (b"d1".to_vec(), b"dd11".to_vec()),
+            (b"d2".to_vec(), b"dd22".to_vec()),
+        ].into_iter()
+            .map(|(k, v)| Some((k, v)));
+        expect_multi_values(
+            results.clone().collect(),
+            <Storage<RocksEngine>>::async_snapshot(storage.get_engine(), &ctx)
+                .and_then(move |snapshot| {
+                    <Storage<RocksEngine>>::raw_scan(
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: Why not invoke `storage.async_raw_scan` directly here, just like other tests here does?

File: src/storage/mod.rs [hunk 8]
```diff
@@ -2491,29 +2519,111 @@ mod tests {
...
+                        false,
+                    )
+                })
+                .wait(),
+        );
+        expect_multi_values(
+            results.rev().collect(),
+            <Storage<RocksEngine>>::async_snapshot(storage.get_engine(), &ctx)
+                .and_then(move |snapshot| {
+                    <Storage<RocksEngine>>::raw_scan(
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: ditto

File: src/storage/mod.rs (hunk 7 above)
Comment (kamijin-fanta) [reply src/storage/mod.rs]: async_raw_scan has no end_key argument. Pass `None` internally.
I wanted to test it.

Comment (MyonKeminta) [reply src/storage/mod.rs]: Oh I see. Please explain it more clearly in the comments.

File: src/storage/mod.rs [hunk 9]
```diff
@@ -1134,14 +1134,28 @@ impl<E: Engine> Storage<E> {
         limit: usize,
         statistics: &mut Statistics,
         key_only: bool,
+        reverse: bool,
     ) -> Result<Vec<Result<KvPair>>> {
         let mut option = IterOption::default();
         if let Some(end) = end_key {
-            option.set_upper_bound(end.into_encoded());
+            if reverse {
```
Comment (breeswish) [inline src/storage/mod.rs]: I am not very enjoy `if reverse` every where here, which is not good in both looking and performance. Do you have some better ideas? @MyonKeminta How about split into two functions?

Comment (MyonKeminta) [reply src/storage/mod.rs]: Good idea

Comment (kamijin-fanta) [reply src/storage/mod.rs]: should I define raw_scan_reverse?

Comment (MyonKeminta) [reply src/storage/mod.rs]: Maybe `reverse_raw_scan`, I think

File: src/storage/mod.rs (hunk 7 above)
Comment (kamijin-fanta) [reply src/storage/mod.rs]: add comments

File: src/storage/mod.rs (hunk 9 above)
Comment (kamijin-fanta) [reply src/storage/mod.rs]: split raw_scan function

File: src/storage/mod.rs [hunk 10]
```diff
@@ -2865,7 +3058,74 @@ mod tests {
...
+            Some((b"a".to_vec(), b"aa".to_vec())),
+            Some((b"b2".to_vec(), b"bb22".to_vec())),
+            Some((b"b1".to_vec(), b"bb11".to_vec())),
+            Some((b"b".to_vec(), b"bb".to_vec())),
+            Some((b"c2".to_vec(), b"cc22".to_vec())),
+            Some((b"c1".to_vec(), b"cc11".to_vec())),
+            Some((b"c".to_vec(), b"cc".to_vec())),
+        ];
+        let ranges: Vec<KeyRange> = vec![
+            (b"a3".to_vec(), b"a".to_vec()),
```
Comment (breeswish) [inline src/storage/mod.rs]: Could you also add a test to cover the scenario that end key is empty (which is checked in `check_key_ranges`)?

File: Cargo.lock [hunk 11]
```diff
@@ -610,7 +610,7 @@ dependencies = [
 [[package]]
 name = "kvproto"
 version = "0.0.1"
-source = "git+https://github.com/pingcap/kvproto.git#1b5d69cd1d263f9473f7ed2847a7660af394c437"
+source = "git+https://github.com/pingcap/kvproto.git#0ba3ca8a6e3775accbe4bdf844162cd7685dafe2"
```
Comment (MyonKeminta) [inline Cargo.lock]: The master branch now uses a newer revision of kvproto (updated by another PR yesterday), and your changes to kvproto is already available in TiKV's master. Please keep this line unchanged, then your PR can be built.

File: src/storage/mod.rs [hunk 12]
```diff
@@ -1224,15 +1269,21 @@ impl<E: Engine> Storage<E> {
...
+    fn check_key_ranges(ranges: &[KeyRange], reverse: bool) -> bool {
         let ranges_len = ranges.len();
         for i in 0..ranges_len {
             let start_key = ranges[i].get_start_key();
             let mut end_key = ranges[i].get_end_key();
             if end_key.is_empty() && i + 1 != ranges_len {
                 end_key = ranges[i + 1].get_start_key();
             }
-            if !end_key.is_empty() && start_key >= end_key {
+            if !end_key.is_empty()
```
[the remark this answers is not in the document]
Comment (kamijin-fanta) [reply src/storage/mod.rs]: When conscious of time series metrics, I feel it naturally. However, I follow the policy of the project.

File: src/storage/mod.rs (hunk 10 above)
Comment (MyonKeminta) [reply src/storage/mod.rs]: Also it will be better if there's a test for the `check_key_ranges` function

File: components/test_storage/src/sync_storage.rs [hunk 13]
```diff
@@ -215,9 +215,10 @@ impl<E: Engine> SyncStorage<E> {
         cf: String,
         start_key: Vec<u8>,
         limit: usize,
+        reverse: bool,
     ) -> Result<Vec<Result<KvPair>>> {
         self.store
-            .async_raw_scan(ctx, cf, start_key, limit, false)
+            .async_raw_scan(ctx, cf, start_key, limit, false, reverse)
```
[the remark this answers is not in the document]
Comment (kamijin-fanta) [reply components/test_storage/src/sync_storage.rs]: done

File: src/storage/mod.rs (hunk 10 above)
Comment (kamijin-fanta) [reply src/storage/mod.rs]: add test case for check_key_ranges

**Comment by breeswish** [comment]
For the following two ranges, it looks wired in the reverse mode.. The first one is a, b, c and the second one is c, b, a. I prefer to keep the same global total order for all possible ranges in reverse mode, i.e. for every range (a_i, b_i) at index i, we have a_i > a_i+1.

```rust
        let ranges = make_ranges(vec![
            (b"a3".to_vec(), b"a".to_vec()),
            (b"b3".to_vec(), b"b".to_vec()),
            (b"c3".to_vec(), b"c".to_vec()),
        ]);
```

```rust
         let ranges = make_ranges(vec![
            (b"c3".to_vec(), vec![]),
            (b"b3".to_vec(), vec![]),
            (b"a3".to_vec(), vec![]),
        ]);
```

**Comment by breeswish** [comment]
How do you think @MyonKeminta

**Comment by kamijin-fanta** [comment]
@breeswish yes

**Comment by kamijin-fanta** [comment]
what is the status of this PR?

**Comment by MyonKeminta** [comment]
Oh, sorry
/run-integration-tests

**Comment by kamijin-fanta** [comment]
Thank you for the merge! :tada:
